Arch packages implementation

ExplodingDragon opened the pull request:
close #25037

This Commit was created by d1nch8g (#25396).

This PR adds a package registry for [Arch Linux](https://archlinux.org/) packages with support for package files, [signatures](https://wiki.archlinux.org/title/Pacman/Package_signing), and automatic [pacman-database](https://archlinux.org/pacman/repo-add.8.html) management.

Features:

1. Push any ` tar.zst ` package and Gitea sign it.
2. Delete endpoint for specific package version and all related files
3. Supports trust levels with `SigLevel = Required`.
4. Package UI with instructions to connect to the new pacman database and visualised package metadata

![](https://github.com/user-attachments/assets/5b734831-9f57-4c26-abf2-6b53981ec1b2)

You can follow [this tutorial](https://wiki.archlinux.org/title/Creating_packages) to build a *.pkg.tar.zst package for testing

docs pr: [gitea/docs#47](https://gitea.com/gitea/docs/pulls/47)

Co-authored-by: d1nch8g@ion.lc
Co-authored-by: @KN4CK3R
Co-authored-by: @silverwind
Co-authored-by: @mahlzahn

ExplodingDragon commented:
Package signatures should be issued by Gitea; custom signatures will result in inconsistencies with index signatures.

KN4CK3R commented:
When #25396 was closed some weeks ago, I adopted the PR and rewrote it to look similar to how the other package registries work. I will push my changes next week, then we could compare with this PR and decide how we continue.

ExplodingDragon commented:
> When #25396 was closed some weeks ago, I adopted the PR and rewrote it to look similar to how the other package registries work. I will push my changes next week, then we could compare with this PR and decide how we continue.

@KN4CK3R My implementation is not perfect, in fact, I just refactored the original pr and delegated its signature to Gitea to manage, like #27069.

ExplodingDragon commented:
> When #25396 was closed some weeks ago, I adopted the PR and rewrote it to look similar to how the other package registries work. I will push my changes next week, then we could compare with this PR and decide how we continue.

@KN4CK3R Any new changes?

KN4CK3R commented:
You can find my code here: https://github.com/go-gitea/gitea/compare/main...KN4CK3R:gitea:feature-arch

techknowlogick commented:
Thanks again @ExplodingDragon for your work on getting these additional package PRs in. I'm going through the backlog of PRs now, and would love to get to this one. Are you able to incorporate @KN4CK3R's work into this PR?

ExplodingDragon commented:
> Thanks again @ExplodingDragon for your work on getting these additional package PRs in. I'm going through the backlog of PRs now, and would love to get to this one. Are you able to incorporate @KN4CK3R's work into this PR?

@techknowlogick This pull request is complete and ready for review at any time.

lunny commented:
@ExplodingDragon Please send a doc pull request in https://gitea.com/gitea/docs like other package documentation. And that PR will be reviewed and merged after this one is merged.

ExplodingDragon commented:
> @ExplodingDragon Please send a doc pull request in https://gitea.com/gitea/docs like other package documentation. And that PR will be reviewed and merged after this one is merged.

@lunny  docs pull request is at https://gitea.com/gitea/docs/pulls/47.

ExplodingDragon commented:
The current pull request contains errors, and the fix patch in [forgejo/forgejo#4903](https://codeberg.org/forgejo/forgejo/pulls/4903) , provided here for archival purposes only.

lunny commented:
> The current pull request contains errors, and the fix patch in [forgejo/forgejo#4903](https://codeberg.org/forgejo/forgejo/pulls/4903) and [forgejo/forgejo#5121](https://codeberg.org/forgejo/forgejo/pulls/5121).

Can you merge them into this PR?

ExplodingDragon commented:
> Can you merge them into this PR?

@lunny  I will merge it over the weekend; I haven't had much time lately.

ExplodingDragon commented:
This PR (Arch packages implementation)  has been completed, Requesting review now.

Ping @KN4CK3R

yp05327 commented on modules/setting/packages.go:
```diff
@@ -79,6 +80,8 @@ func loadPackagesFrom(rootCfg ConfigProvider) (err error) {
 
 	Packages.LimitTotalOwnerSize = mustBytes(sec, "LIMIT_TOTAL_OWNER_SIZE")
 	Packages.LimitSizeAlpine = mustBytes(sec, "LIMIT_SIZE_ALPINE")
+	Packages.LimitSizeArch = mustBytes(sec, "LIMIT_SIZE_ARCH")
+	Packages.LimitSizeArch = mustBytes(sec, "LIMIT_SIZE_ARCH")
```
？？？

yp05327 commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,255 @@
+	})
+	if err != nil {
+		apiError(ctx, http.StatusInternalServerError, err)
+	}
```
return?

yp05327 commented:
![image](https://github.com/user-attachments/assets/21235c96-2b75-4654-bb60-e9be263064fd)

lunny commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,143 @@
+		</div>
+		<div class="field">
+			<label>{{ctx.Locale.Tr "packages.registry.documentation" "Arch"
+				"https://forgejo.org/docs/latest/user/packages/arch/"}}</label>
```
The documentation link is not right.

lunny commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,346 @@
+		case ".PKGINFO":
+			pkg, err = ParsePackageInfo(tarballType, f)
+			if err != nil {
+				return nil, err
```
f.Close should be invoked here.

lunny commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,346 @@
+			continue
+		}
+
+		key, value, find := strings.Cut(line, "=")
```
I don't know whether there will be a line like `a=b #this is comment`.

ExplodingDragon replied:
> I don't know whether there will be a line like `a=b #this is comment`.

@lunny No, this file is automatically generated and this will not happen.

lunny commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,257 @@
+
+func PushPackage(ctx *context.Context) {
+	group := ctx.PathParam("group")
+
```
I think we also need to consider the parallel uploading problem. ref https://github.com/go-gitea/gitea/pull/31954/files

lunny commented on services/packages/arch/repository.go:
```diff
@@ -0,0 +1,364 @@
+		for _, ver := range versions {
+			files, err := packages_model.GetFilesByVersionID(ctx, ver.ID)
+			if err != nil {
+				return nil, errors.Join(tw.Close(), gw.Close(), db.Close(), err)
```
Should we really need to return err from `tw.Close` and others? Maybe move them all to a defer function?

yp05327 commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,347 @@
+			return util.NewInvalidArgumentErrorf("invalid replaces: %s", p)
+		}
+	}
+	for _, p := range p.VersionMetadata.Replaces {
```
```suggestion
	for _, x := range p.VersionMetadata.Xdata{
```
And above also has some copy paste for the param name of for loop.

yp05327 replied:
But I don't prefer using same param name in for loop and outside it.

lunny commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,273 @@
+}
+
+func refreshLocker(ctx *context.Context, group string) (globallock.ReleaseFunc, error) {
+	return globallock.Lock(ctx, fmt.Sprintf("pkg_arch_pkg_%s", group))
```
Is this related to some user/org name?

yp05327 commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,342 @@
+	if err != nil {
+		return nil, err
+	}
+	header := make([]byte, 5)
```
maybe add this to vars is better:
`maxMagicLength = 5`

yp05327 replied:
I'm not sure whether this is defined in the protocol. If it is, maybe this is fine.

ExplodingDragon replied:
@yp05327  It is not possible to determine the final sample size, different compression uses different magic values, here `5` is the magic length of `xz`, which may change subsequently.

ExplodingDragon replied:
In fact this is a lazy scheme where the user could forge similar packages to pass the authentication, but this is not something we consider as this behavior will fail subsequently.

yp05327 replied:
This may cause a risk that maintainer may not notice the init of `header`, and only added new magics below.
Then `out of range` panic may happen.

yp05327 commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,273 @@
+	}
+	deleted := false
+	for _, file := range files {
+		extName := fmt.Sprintf("-%s.pkg.tar%s", pkgArch, filepath.Ext(file.LowerName))
```
Is this correct? 
`-%s.pkg.tar%s`
or
`-%s.pkg.tar.%s`?
Also some places has this strange ext

ExplodingDragon replied:
@yp05327 That's correct, `tar%s` is calculated by `filepath.Ext` and `tar.%s` is a constant.

yp05327 commented on services/packages/arch/repository.go:
```diff
@@ -0,0 +1,372 @@
+	var (
+		splt    = strings.Split(file, "-")
+		pkgname = strings.Join(splt[0:len(splt)-3], "-")
+		vername = splt[len(splt)-3] + "-" + splt[len(splt)-2]
```
How to ensure `len(splt) > 3`

yp05327 commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,143 @@
+
+{{end -}}
+{{- $GroupSize := (len .Groups) -}}
+{{-  range $i,$v :=  .Groups -}}
```
```suggestion
{{- range $i,$v := .Groups -}}
```

yp05327 commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,143 @@
+				<pre
+					class="code-block"><code>
+{{- if gt (len $.Groups) 1 -}}
+# {{ctx.Locale.Tr "packages.arch.pacman.repo.multi"  $.PackageDescriptor.Package.LowerName}}
```
```suggestion
# {{ctx.Locale.Tr "packages.arch.pacman.repo.multi" $.PackageDescriptor.Package.LowerName}}
```

ExplodingDragon commented:
E2E Testing:

TODO

yp05327 commented on templates/package/content/arch.tmpl:
```diff
@@ -0,0 +1,143 @@
+				<td>{{.PackageDescriptor.Metadata.Description}}</td>
+			</tr>
+
+			{{if .PackageDescriptor.Metadata.Groups}}
+			<tr>
+				<td class="collapsing">
+					<h5>{{ctx.Locale.Tr "packages.arch.version.groups"}}</h5>
+				</td>
+				<td>{{StringUtils.Join $.PackageDescriptor.Metadata.Groups ", "}}</td>
+			</tr>
+			{{end}}
```
Maybe this can be simplified?

```
$locale=""
$value=""
if xxx {
  $locale=ctx.Locale.Tr xxx
  $value=xxxx
}
if ....
<HTML code once>
```

ExplodingDragon replied:
@yp05327  There is no need; using `variables + if` doesn't seem as intuitive as the current code.

yp05327 replied:
If I want to add `xxx` class, then I need to add `xxx` class 9 times.

yp05327 replied:
ah, should have a `range`. 🤔

KN4CK3R reviewed (changes requested):
Blocking until I have time to have a look.

wxiaoguang commented:
> Blocking until I have time to have a look.

Do you have a estimated time? 1.23 is releasing soon?

wxiaoguang commented:
Made some changes:
* Avoid using global variables (it slows down non-web commands).
* Avoid bloating api.go, make the handlers process the requests.
* Use unified PackageRegistryHost to avoid duplicate code.

Last call: if there is no other problem, I think this PR could be merged in a few days since it has enough tests.

delvh commented:
@KN4CK3R did you have time to take a look at this PR?

wxiaoguang commented on tests/integration/api_packages_arch_test.go:
```diff
@@ -336,6 +336,30 @@ HMhNSS1IzUsBcpJAPFAwwUXSM0u4BjoaR8EoGAWjgGQAAILFeyQADAAA
 			MakeRequest(t, req, http.StatusNoContent)
 		}
 	})
+	t.Run("Package Arch Test", func(t *testing.T) {
```
The test name seems strange (indeed the while file is for "package arch test"), will improve it.

KN4CK3R commented on services/packages/arch/repository.go:
```diff
@@ -0,0 +1,370 @@
+
+// GetPackageFile Get data related to provided filename and distribution, for package files
+// update download counter.
+func GetPackageFile(ctx context.Context, group, file string, ownerID int64) (io.ReadSeekCloser, *url.URL, *packages_model.PackageFile, error) {
```
Why not use one of the existing methods in `code.gitea.io/gitea/services/packages`?

wxiaoguang replied:
It already uses `services/packages`, but here it needs to handle the filenames first. 

Do you have some better `services/packages` functions to call here?

KN4CK3R commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,387 @@
+}
+
+// Desc Create pacman package description file.
+func (p *Package) Desc() string {
```
I would move this code into the `repository.go` file to the other db related logic. Then it's not needed to have the `MD5SUM` and `SHA256SUM` fields because that info is already present.

KN4CK3R commented on services/packages/arch/repository.go:
```diff
@@ -0,0 +1,370 @@
+	return packages_service.GetPackageFileStream(ctx, pkgFile)
+}
+
+func GetPackageDBFile(ctx context.Context, group, arch string, ownerID int64, signFile bool) (io.ReadSeekCloser, *url.URL, *packages_model.PackageFile, error) {
```
Same here. I think this should be handled by the router.

KN4CK3R commented on tests/integration/api_packages_arch_test.go:
```diff
@@ -0,0 +1,435 @@
+	defer tests.PrepareTestEnv(t)()
+	user := unittest.AssertExistsAndLoadBean(t, &user_model.User{ID: 2})
+	unpack := func(s string) []byte {
+		data, _ := base64.StdEncoding.DecodeString(strings.ReplaceAll(strings.ReplaceAll(strings.TrimSpace(s), "\n", ""), "\r", ""))
```
All these string replacements are not necessary for `base64.StdEncoding.DecodeString`.

https://github.com/go-gitea/gitea/blob/331e878e81d57235a53199383087c7649797a887/tests/integration/api_packages_alpine_test.go#L60C18-L60C50

wxiaoguang replied:
26c06a365a7141dd9ee8df2a70cd54cca79c39f7

KN4CK3R commented on tests/integration/api_packages_arch_test.go:
```diff
@@ -0,0 +1,435 @@
+	}
+	rootURL := fmt.Sprintf("/api/packages/%s/arch", user.Name)
+
+	pkgs := map[string][]byte{
```
If we need so many test packages it may be better to generate them on the fly.

wxiaoguang replied:
I have added comments to them, we could improve them later (I also have done so, for example #31357)

KN4CK3R commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,387 @@
+		magic    []byte
+		archiver func() archiver.Reader
+	}{
+		"zst": {
+			magic: []byte{0x28, 0xB5, 0x2F, 0xFD},
+			archiver: func() archiver.Reader {
+				return archiver.NewTarZstd()
+			},
+		},
+		"xz": {
+			magic: []byte{0xFD, 0x37, 0x7A, 0x58, 0x5A},
+			archiver: func() archiver.Reader {
+				return archiver.NewTarXz()
+			},
+		},
+		"gz": {
+			magic: []byte{0x1F, 0x8B},
+			archiver: func() archiver.Reader {
+				return archiver.NewTarGz()
+			},
+		},
```
Are there any other containers than zst in the wild? All repositories I have tested just contain zst.

KN4CK3R commented on routers/api/packages/arch/arch.go:
```diff
@@ -0,0 +1,284 @@
+}
+
+func GetPackageOrDB(ctx *context.Context) {
+	pathFields := strings.Split(strings.Trim(ctx.PathParam("*"), "/"), "/")
+	pathFieldsLen := len(pathFields)
+	if pathFieldsLen < 2 {
+		ctx.Status(http.StatusBadRequest)
+		return
+	}
+
+	group := strings.Join(pathFields[:pathFieldsLen-2], "/")
+	arch := pathFields[pathFieldsLen-2]
+	file := pathFields[pathFieldsLen-1]
```
Why not simply specify a clean route with parameters so you don't need to rebuild the logic here?

wxiaoguang commented:
> I still think this PR could be simpler just by following how the other package types are structured.
> 
> My implementation can still be found here: [main...KN4CK3R:gitea:feature-arch](https://github.com/go-gitea/gitea/compare/main...KN4CK3R:gitea:feature-arch)

If we'd like to include it in 1.23, what's your suggestion to move one?

I think we have 3 choices:

1. Merge this as-is and refactor it later
2. You help to improve this PR and merge it
3. Use your implementation in a new PR and close this one

Personally I think 1 could work.

What's your opinion?

KN4CK3R commented:
Problem with option 1 is that the refactoring would most likely use different routes and perhaps different metadata (stored as json) which may require migrations. So it's 2 or 3 for me. @ExplodingDragon are you willing to make/accept changes?

ExplodingDragon commented:
@KN4CK3R 

> Are there any other containers than zst in the wild? 

Yes, archlinuxarm uses xz.

>  use different routes and perhaps different metadata (stored as json) 

The use of non-json storage is to avoid serialization when generating the db.

> So it's 2 or 3 for me. @ExplodingDragon are you willing to make/accept changes?

I'm ok with both, 3 is better for me, I don't have much time to work on things, I'll close this pr later, other things are up to you to determine.

ExplodingDragon commented:
Thanks @wxiaoguang and @yp05327 , sorry.

wxiaoguang commented:
@KN4CK3R what's the next plan? Would you like to work on this PR, or propose a new one?

lunny commented:
I think the main conflict is the router path, other codes could be refactored later. Maybe we can do a discussion about which change's routers are better?

wxiaoguang commented:
> @KN4CK3R what's the next plan? Would you like to work on this PR, or propose a new one?

If no plan, then still merge this as-is?

wxiaoguang commented:
Did some tests:

> This PR:

```
    testlogger.go:62: 2024/11/17 20:43:00 ...eb/routing/logger.go:102:func1() [I] router: completed GET /api/packages/user2/arch/x86_64/test-1.0.0-1-x86_64.pkg.tar.zst for test-mock:12345, 200 OK in 1.1ms @ arch/arch.go:177(arch.GetPackageOrDB)
    testlogger.go:62: 2024/11/17 20:43:00 ...eb/routing/logger.go:102:func1() [I] router: completed GET /api/packages/user2/arch/x86_64/test-1.0.0-1-any.pkg.tar.zst for test-mock:12345, 200 OK in 0.9ms @ arch/arch.go:177(arch.GetPackageOrDB)
    testlogger.go:62: 2024/11/17 20:43:00 ...eb/routing/logger.go:102:func1() [I] router: completed GET /api/packages/user2/arch/unknown/x86_64/test-1.0.0-1-aarch64.pkg.tar.zst for test-mock:12345, 404 Not Found in 0.5ms @ arch/arch.go:177(arch.GetPackageOrDB)

    testlogger.go:62: 2024/11/17 20:43:00 ...eb/routing/logger.go:102:func1() [I] router: completed GET /api/packages/user2/arch/arch/x86_64/test-1.0.0-1-x86_64.pkg.tar.zst for test-mock:12345, 200 OK in 1.1ms @ arch/arch.go:177(arch.GetPackageOrDB)
    testlogger.go:62: 2024/11/17 20:43:00 ...eb/routing/logger.go:102:func1() [I] router: completed GET /api/packages/user2/arch/arch/x86_64/test-1.0.0-1-any.pkg.tar.zst for test-mock:12345, 200 OK in 1.0ms @ arch/arch.go:177(arch.GetPackageOrDB)
    testlogger.go:62: 2024/11/17 20:43:00 ...eb/routing/logger.go:102:func1() [I] router: completed GET /api/packages/user2/arch/unknown/x86_64/test-1.0.0-1-aarch64.pkg.tar.zst for test-mock:12345, 404 Not Found in 0.5ms @ arch/arch.go:177(arch.GetPackageOrDB)

```

> [main...KN4CK3R:gitea:feature-arch](https://github.com/go-gitea/gitea/compare/main...KN4CK3R:gitea:feature-arch)

```
    testlogger.go:60: 2024/11/17 20:40:48 ...eb/routing/logger.go:102:func1() [I] router: completed PUT /api/packages/user2/arch/main for test-mock:12345, 201 Created in 15.9ms @ arch/arch.go:44(arch.UploadPackageFile)
    testlogger.go:60: 2024/11/17 20:40:48 ...eb/routing/logger.go:102:func1() [I] router: completed GET /api/packages/user2/arch/main/aarch64/packages.db for test-mock:12345, 200 OK in 0.8ms @ arch/arch.go:147(arch.DownloadPackageOrRepositoryFile)
    testlogger.go:60: 2024/11/17 20:40:48 ...eb/routing/logger.go:102:func1() [I] router: completed GET /api/packages/user2/arch/main/any/gitea-test_any-1.4.1-r3-any.pck.tar.zst for test-mock:12345, 200 OK in 1.0ms @ arch/arch.go:147(arch.DownloadPackageOrRepositoryFile)
    testlogger.go:60: 2024/11/17 20:40:48 ...eb/routing/logger.go:102:func1() [I] router: completed GET /api/packages/user2/arch/main/aarch64/gitea-test_any-1.4.1-r3-any.pck.tar.zst for test-mock:12345, 200 OK in 1.3ms @ arch/arch.go:147(arch.DownloadPackageOrRepositoryFile)
    testlogger.go:60: 2024/11/17 20:40:48 ...eb/routing/logger.go:102:func1() [I] router: completed GET /api/packages/user2/arch/main/myarch/gitea-test_any-1.4.1-r3-any.pck.tar.zst for test-mock:12345, 200 OK in 1.1ms @ arch/arch.go:147(arch.DownloadPackageOrRepositoryFile)
    testlogger.go:60: 2024/11/17 20:40:48 ...eb/routing/logger.go:102:func1() [I] router: completed DELETE /api/packages/user2/arch/main/any/gitea-test_any-1.4.1-r3-any.pck.tar.zst for test-mock:12345, 204 No Content in 7.7ms @ arch/arch.go:233(arch.DeletePackageFile)
```


We can see the design is different: there is no `{repository}` part in this PR (the repository is 'main' or 'testing' in KN4CK3R's code)

wxiaoguang commented:
Let's start a fresh discussion @ExplodingDragon @KN4CK3R 

----

I have made the tests in this PR pass KN4CK3R 's code: https://github.com/wxiaoguang/gitea/commits/feature-arch-KN4CK3R/ , see the last commit (and I might still force push the last commit a few times to make it easier to read)

I marked the differences with `// DIFF:`, they are (this VS KN4CK3R's):

1. `DIFF:RepositoryKey`:  `/repository.key` VS `/key`
1. `DIFF:RepositoryName`: this PR supports empty repo name and slashes in repo name. KN4CK3R's only supports a valid symbol name for repo name (it answers KN4CK3R's question: why the path handling is more complex here)
1. `DIFF:InvalidPackage`: server response code 40x VS 500
1. `DIFF:PackageFileStore`: `sig` file in package files VS not in
1. `DIFF:PackageFileNameExt`: `pkg` VS `pck`
1. `DIFF:RepositoryDBFiles`: `[..../desc]` VS `[.../files, ..../desc]`
1. `DIFF:DeletePackageURL`: DELETE `/test/1.0.0-1/x86_64` VS `/x86_64/test-1.0.0-1-x86_64.pck.tar.zst`
1. `DIFF:OtherPackageType`: this PR supports xz and gz, KN4CK3R's doesn't

So what implementation should we take? Some differences are quite important (eg: DeletePackageURL)

wxiaoguang commented on modules/packages/arch/metadata.go:
```diff
@@ -0,0 +1,387 @@
+}
+
+// ValidatePackageSpec Arch package validation according to PKGBUILD specification.
+func ValidatePackageSpec(p *Package) error {
```
26c06a365a7141dd9ee8df2a70cd54cca79c39f7

wxiaoguang commented:
Overall I think this PR is good enough, if the differences in https://github.com/go-gitea/gitea/pull/31037#issuecomment-2481280902 don't block.

wxiaoguang commented:
@KN4CK3R last call if the differences in https://github.com/go-gitea/gitea/pull/31037#issuecomment-2481280902 don't matter.

KN4CK3R commented:
I will push some changes later.

KN4CK3R commented:
I pushed the changes (coauthored by @ExplodingDragon) some days ago and removed the docs file. Maybe someone would like to have a look:
https://github.com/go-gitea/gitea/compare/main...KN4CK3R:gitea:feature-arch

wxiaoguang commented:
Thank you, are there some details about the DIFF https://github.com/go-gitea/gitea/pull/31037#issuecomment-2481280902  ?

> DIFF:RepositoryName: this PR supports empty repo name and slashes in repo name. KN4CK3R's only supports a valid symbol name for repo name (it answers KN4CK3R's question: why the path handling is more complex here)

ExplodingDragon's support empty repo name, the new one doesn't support? (To be clear, whether it should be supported?)

> DIFF:DeletePackageURL: DELETE /test/1.0.0-1/x86_64 VS /x86_64/test-1.0.0-1-x86_64.pck.tar.zst

Which URL is correct (or better) to delete?

----

And one more thing, in your `routers/api/packages/api.go`, you bloated the api.go again ... I have managed to move these complex logic into handlers in 84ff3538045b24af67059f417869b116591bde7a , I do not think it is right to write a lot of code in `api.go`

wxiaoguang commented:
> > Which URL is correct (or better) to delete?
> 
> There is no correct because there is no official spec. Deleting a package should rarely happen. It may be easier for the user to just have the package name and version in the url instead of crafting the full filename. Providing the filename makes the code easier. Removing this endpoint may be even easier.


`DELETE /test/1.0.0-1/x86_64` looks clearer to me.

The question is: does the extension name matter? If the extension name matters, then it must use `DELETE /x86_64/test-1.0.0-1-x86_64.pkg.tar.zst` (but, why `x86_64` duplicates?). Otherwise, `DELETE /test/1.0.0-1/x86_64` seems more general and works for all extensions.




> > you bloated the api.go again
> 
> The only reason for this code is because chi doesn't support `/*/{param}`. Therefore it's chi related and belongs into the api.go.

I do not quite agree, we can't blame `chi` everytime and keep adding more code to the router init function, it definitely will make the `api.go` become a monster and more and more difficult to maintain.

(That's my option and I won't block it.)

wxiaoguang commented:
And one more question, how to continue? Would you like to push your changes here, or create a new PR?

wxiaoguang commented:
@KN4CK3R maybe 1.23 is going to be frozen in a week. The last 2 questions:

* Which one should be used to "DELETE"?
    * `DELETE /.../{repo-group}/test/1.0.0-1/x86_64`: simple and no extension name (ExplodingDragon's)
    * `DELETE /.../{repo-group}/x86_64/test-1.0.0-1-x86_64.pkg.tar.zst`: have full file name and duplicate architecture (KN4CK3R's)
    * `DELETE /.../{repo-group}/test-1.0.0-1-x86_64.pkg.tar.zst`: only full file name, no duplicate architecture?
* How to continue?
    * push KN4CK3R's changes here
    * create a new PR from KN4CK3R's branch?

KN4CK3R commented:
I already updated the delete endpoint in my code. If @ExplodingDragon is fine with it, I would like to create a PR.

wxiaoguang commented:
> I already updated the delete endpoint in my code. If @ExplodingDragon is fine with it, I would like to create a PR.

According to https://github.com/go-gitea/gitea/pull/31037#issuecomment-2466055941

> > I'm ok with both, 3 is better for me, I don't have much time to work on things, I'll close this pr later, other things are up to you to determine.

🙏

ExplodingDragon commented:
This pr has been implemented by @KN4CK3R.

wxiaoguang commented:
> > And one more thing, in your `routers/api/packages/api.go`, you bloated the api.go again ... I have managed to move these complex logic into handlers in [84ff353](https://github.com/go-gitea/gitea/commit/84ff3538045b24af67059f417869b116591bde7a) , I do not think it is right to write a lot of code in `api.go`
>
> The only reason for this code is because chi doesn't support `/*/{param}`. Therefore it's chi related and belongs into the api.go.

To make the routes clearer:  Refactor arch route handlers #32972 (the first step) , I am sure we could have a quite clear router init function soon.
